tests/src/Interop/ICustomMarshaler/ICustomMarshaler.cs

```diff
@@ -0,0 +1,579 @@
+            public object MarshalNativeToManaged(IntPtr pNativeData) => null;
+
+            public static ICustomMarshaler GetInstance(string cookie) => new StringForwardingCustomMarshaler();
+        }
+
+        [DllImport(LibcLibrary, EntryPoint = "atoi")]
+        public static extern int MarshalerOnStringTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(StringForwardingCustomMarshaler))] string str);
+
+        [Fact]
+        public void CustomMarshaler_ArrayType_Success()
+        {
+            Assert.Equal(123, MarshalerOnArrayTypeMethod(new string[] { "123" }));
+        }
+
+        public class ArrayForwardingCustomMarshaler : ICustomMarshaler
+        {
+            public void CleanUpManagedData(object ManagedObj) { }
+            public void CleanUpNativeData(IntPtr pNativeData) { Marshal.ZeroFreeCoTaskMemAnsi(pNativeData); }
+
+            public int GetNativeDataSize() => IntPtr.Size;
+
+            public IntPtr MarshalManagedToNative(object ManagedObj) => Marshal.StringToCoTaskMemAnsi(((string[])ManagedObj)[0]);
+            public object MarshalNativeToManaged(IntPtr pNativeData) => null;
+
+            public static ICustomMarshaler GetInstance(string cookie) => new ArrayForwardingCustomMarshaler();
+        }
+
+        [DllImport(LibcLibrary, EntryPoint = "atoi")]
+        public static extern int MarshalerOnArrayTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalType = "System.Runtime.InteropServices.Tests.ICustomMarshalerTests+ArrayForwardingCustomMarshaler")] string[] str);
+
+        [Fact]
+        public void CustomMarshaler_BoxedValueType_Success()
+        {
+            Assert.Equal(246, MarshalerOnBoxedValueTypeMethod(123));
+        }
+
+        public class BoxedValueTypeCustomMarshaler : ICustomMarshaler
+        {
+            public void CleanUpManagedData(object ManagedObj) { }
+            public void CleanUpNativeData(IntPtr pNativeData) { Marshal.ZeroFreeCoTaskMemAnsi(pNativeData); }
+
+            public int GetNativeDataSize() => IntPtr.Size;
+
+            public IntPtr MarshalManagedToNative(object ManagedObj)
+            {
+                int unboxedValueType = (int)ManagedObj * 2;
+                return Marshal.StringToCoTaskMemAnsi(unboxedValueType.ToString());
+            }
+
+            public object MarshalNativeToManaged(IntPtr pNativeData) => null;
+
+            public static ICustomMarshaler GetInstance(string cookie) => new BoxedValueTypeCustomMarshaler();
+        }
+
+        [DllImport(LibcLibrary, EntryPoint = "atoi")]
+        public static extern int MarshalerOnBoxedValueTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(BoxedValueTypeCustomMarshaler))] object i);
+
+        [Fact]
+        public void Parameter_CustomMarshalerProvidedOnClassType_ForwardsCorrectly()
+        {
+            Assert.Equal("246", MarshalerOnClassTypeMethod(new StringContainer { Value = "123" }).Value);
```

Comment:
This test fails on my mac but passes in a Windows VM - anyone got ideas? Could be invoking UB - is it a bad/wrong test?

Reply:
This test seems okay. However, I think `ClassForwardingCustomMarshaler.CleanUpNativeData()` is wrong. I am getting two calls to that function (1) the string and then (2) the `IntPtr` that is passed to `MarshalNativeToManaged()`. I think this works on Windows because of the Win32Atom contract and your input - `123` - is properly ignored. Try something > 64k and I bet it fails everywhere.